File: src/main/GO/leetcode/lists/hot100/LeetCode124.go
package hot100

import "math"

func maxPathSum(root *TreeNode) int {
	if root == nil {
		return 0
	}
	max := math.MinInt32
	var dfs func(root *TreeNode) int
	dfs = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		l := Max(dfs(root.Left), 0)
		r := Max(dfs(root.Right), 0)
		max = Max(max, l+r+root.Val)
		return Max(l, r) + root.Val
	}
	dfs(root)
	return max
}

File: src/main/java/leetcode/lists/hot100/LeetCode124.java
package leetcode.lists.hot100;

import common.*;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/5 11:36
 */
public class LeetCode124 {
    int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }
        maxGain(root);
        return maxSum;
    }

    private int maxGain(TreeNode node) {
        if (node == null) {
            return 0;
        }
        //每个节点的最大贡献值
        int left = Math.max(maxGain(node.left), 0);
        int right = Math.max(maxGain(node.right), 0);

        int path = node.val + left + right;
        //得到二叉树的最大路径和
        maxSum = Math.max(maxSum, path);
        // 返回节点的最大贡献值
        return node.val + Math.max(left, right);
    }
}
